ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -349,51 +352,48 @@ private boolean updateBlobMetadata(BlobId blobId, String fieldName, Object value
   }
 
   @Override
-  public boolean purgeBlob(CloudBlobMetadata blobMetadata) throws CloudStorageException {
-    String blobId = blobMetadata.getId();
-    String blobFileName = azureBlobDataAccessor.getAzureBlobName(blobMetadata);
-    String containerName = azureBlobDataAccessor.getAzureContainerName(blobMetadata);
-    String partitionPath = blobMetadata.getPartitionId();
-    azureMetrics.blobDeleteRequestCount.inc();
+  public int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws CloudStorageException {
+    if (blobMetadataList.isEmpty()) {
+      return 0;
+    }
+    azureMetrics.blobDeleteRequestCount.inc(blobMetadataList.size());
     Timer.Context deleteTimer = azureMetrics.blobDeletionTime.time();
     try {
-      // delete blob from storage
-      boolean deletionDone = azureBlobDataAccessor.deleteFile(containerName, blobFileName);
-
-      // Delete the document too
-      try {
-        cosmosDataAccessor.deleteMetadata(blobMetadata);
-        deletionDone = true;
-        logger.debug("Purged blob {} from partition {}.", blobId, partitionPath);
-      } catch (DocumentClientException dex) {
-        if (dex.getStatusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
-          logger.warn("Could not find metadata for blob {} to delete", blobId);
-        } else {
-          throw dex;
-        }
+      List<CloudBlobMetadata> deletedBlobs = azureBlobDataAccessor.purgeBlobs(blobMetadataList);
+      azureMetrics.blobDeletedCount.inc(deletedBlobs.size());
+      azureMetrics.blobDeleteErrorCount.inc(blobMetadataList.size() - deletedBlobs.size());
+
+      // Remove them from Cosmos too
+      for (CloudBlobMetadata blobMetadata : deletedBlobs) {
+        deleteFromCosmos(blobMetadata);
       }
-      azureMetrics.blobDeletedCount.inc(deletionDone ? 1 : 0);
-      return deletionDone;
-    } catch (Exception e) {
-      azureMetrics.blobDeleteErrorCount.inc();
-      String error = (e instanceof DocumentClientException) ? "Failed to delete metadata document for blob " + blobId
-          : "Failed to delete blob " + blobId + ", storage path: " + containerName + "/" + blobFileName;
-      throw toCloudStorageException(error, e);
+      return deletedBlobs.size();
```

Comment:
I just want to confirm my understanding of the compaction logic: In the case where `deletedBlobs.size() < blobMetadataList.size()`, we rely on the fact that not deleting the record from cosmos will allow the dead blobs query to return the same blob ID in future compaction iterations?

Reply:
Right.